release/src/router/tor/doc/tor-gencert.1.in
'\" t
.\"     Title: tor-gencert
.\"    Author: Nick Mathewson
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 06/08/2017
.\"    Manual: Tor Manual
.\"    Source: Tor
.\"  Language: English
.\"
.TH "TOR\-GENCERT" "1" "06/08/2017" "Tor" "Tor Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
tor-gencert \- Generate certs and keys for Tor directory authorities
.SH "SYNOPSIS"
.sp
\fBtor\-gencert\fR [\-h|\-\-help] [\-v] [\-r|\-\-reuse] [\-\-create\-identity\-key] [\-i \fIid_file\fR] [\-c \fIcert_file\fR] [\-m \fInum\fR] [\-a \fIaddress\fR:\fIport\fR]
.SH "DESCRIPTION"
.sp
\fBtor\-gencert\fR generates certificates and private keys for use by Tor directory authorities running the v3 Tor directory protocol, as used by Tor 0\&.2\&.0 and later\&. If you are not running a directory authority, you don\(cqt need to use tor\-gencert\&.
.sp
Every directory authority has a long term authority \fIidentity\fR \fIkey\fR (which is distinct from the identity key it uses as a Tor server); this key should be kept offline in a secure location\&. It is used to certify shorter\-lived \fIsigning\fR \fIkeys\fR, which are kept online and used by the directory authority to sign votes and consensus documents\&.
.sp
After you use this program to generate a signing key and a certificate, copy those files to the keys subdirectory of your Tor process, and send Tor a SIGHUP signal\&. DO NOT COPY THE IDENTITY KEY\&.
.SH "OPTIONS"
.PP
\fB\-v\fR
.RS 4
Display verbose output\&.
.RE
.PP
\fB\-h\fR or \fB\-\-help\fR
.RS 4
Display help text and exit\&.
.RE
.PP
\fB\-r\fR or \fB\-\-reuse\fR
.RS 4
Generate a new certificate, but not a new signing key\&. This can be used to change the address or lifetime associated with a given key\&.
.RE
.PP
\fB\-\-create\-identity\-key\fR
.RS 4
Generate a new identity key\&. You should only use this option the first time you run tor\-gencert; in the future, you should use the identity key that\(cqs already there\&.
.RE
.PP
\fB\-i\fR \fIFILENAME\fR
.RS 4
Read the identity key from the specified file\&. If the file is not present and \-\-create\-identity\-key is provided, create the identity key in the specified file\&. Default: "\&./authority_identity_key"
.RE
.PP
\fB\-s\fR \fIFILENAME\fR
.RS 4
Write the signing key to the specified file\&. Default: "\&./authority_signing_key"
.RE
.PP
\fB\-c\fR \fIFILENAME\fR
.RS 4
Write the certificate to the specified file\&. Default: "\&./authority_certificate"
.RE
.PP
\fB\-m\fR \fINUM\fR
.RS 4
Number of months that the certificate should be valid\&. Default: 12\&.
.RE
.PP
\fB\-\-passphrase\-fd\fR \fIFILEDES\fR
.RS 4
Filedescriptor to read the passphrase from\&. Ends at the first NUL or newline\&. Default: read from the terminal\&.
.RE
.PP
\fB\-a\fR \fIaddress\fR:\fIport\fR
.RS 4
If provided, advertise the address:port combination as this authority\(cqs preferred directory port in its certificate\&. If the address is a hostname, the hostname is resolved to an IP before it\(cqs published\&.
.RE
.SH "BUGS"
.sp
This probably doesn\(cqt run on Windows\&. That\(cqs not a big issue, since we don\(cqt really want authorities to be running on Windows anyway\&.
.SH "SEE ALSO"
.sp
\fBtor\fR(1)
.sp
See also the "dir\-spec\&.txt" file, distributed with Tor\&.
.SH "AUTHORS"
.sp
.if n \{\
.RS 4
.\}
.nf
Roger Dingledine <arma@mit\&.edu>, Nick Mathewson <nickm@alum\&.mit\&.edu>\&.
.fi
.if n \{\
.RE
.\}
.SH "AUTHOR"
.PP
\fBNick Mathewson\fR
.RS 4
Author.
.RE


Changelog.txt
Asuswrt-Merlin Changelog
========================

380.70 (8-Apr-2018)
  - NOTE: This will be the final 380.xx release for
          all models.  The RT-N66U and RT-AC66U
          support will be dropped, and all other
          models have been migrated to the new gen
          branch, as of release 384.4.

          People who wish to keep getting updates for
          these two older models should look at the
          john9527 fork: https://bit.ly/2EV5Oat

  - CHANGED: Tightened security around some config files.
  - CHANGED: Samba protocol support can now be set to
             SMBv1, SMBv2, or SMBv1 + SMBv2 (the new default).
             This will result in a performance drop on all
             models, but will be more secure.
             Ideally, people should change it to SMBv2 only,
             and then reboot all their client devices to start
             using only the new protocol.
             If performance is more important than security to
             you, then you can switch it back to SMBv1, which is
             the old default behaviour.
  - CHANGED: Switched to the new Entware repo for armv7 models.
             To upgrade, run the following commands TWICE:

             opkg update; opkg upgrade

  - FIXED: Apply button not working on the OpenVPN
           Client page.
  - FIXED: Potential racing condition that could lead to two
           instances of miniupnpd running at boot time.
  - FIXED: Broken FAQ links (backport from 380_8120)
  - FIXED: Security issue in httpd (CVE-2018-8879).
  - FIXED: Security issues in httpd (backports from 380_8228)


380.69_2 (28-Jan-2018)
  - NOTE: The official IRC channel has moved to
          Freenode (#asuswrt).

  - CHANGED: Quantenna watchdog is less likely now to
             incorrectly assume the QTN CPU has crashed
             (which can lead to router reboots). (RT-AC87U)
  - FIXED: IE11 field validation issues on OpenVPN and
           DHCP pages.
  - FIXED: Router crash when importing an OpenVPN certificate
           longer than 3499 characters (the supported limit)
  - FIXED: Users were allowed to enter invalid characters on
           some of the OpenVPN client page fields.
  - FIXED: CVE-2018-5999 in httpd (backport from 384_10007)
  - FIXED: CVE-2018-5721 in httpd (Merlin & theMIROn)


380.69 (11-Dec-2017)
  - NEW: Added option to disable the Asus NAT tunnel service under
         Other Settings -> Tweak.  Not quite sure what this
         partly closed source service is for, but it eats a
         fair amount of CPU and RAM (backport from 382)

  - CHANGED: Updated odhcp6c to be in sync with upstream
             (patch by theMIRon)
  - CHANGED: Updated libogg to 1.3.3 and libvorbis to 1.3.5.
  - CHANGED: Updated wget to 1.19.2 (fixing connectivity to some
             TLS 1.2 servers)
  - CHANGED: Updated RT-N66U and RT-AC66U SDK to GPL 380_8120's
             (fixing KRACK in repeater/bridge mode)
  - CHANGED: Updated openssl to 1.0.2n.
  - CHANGED: Updated tor to 0.2.9.14.
  - FIXED: allow IA_NA mode downgrade with forced IA_PD
           (for ISPs with broken IPv6 support)
           (patch by theMIRon)
  - FIXED: Trend Micro signature check might randomly fail the
           RSA validation.
  - FIXED: Security issues CVE-2017-15275, CVE-2017-12163
           and CVE-2017-12150 (backported to Samba 3.6 and 3.5)
  - FIXED: Httpd crash when accessing certain webui pages with
           no connected Ethernet clients
  - FIXED: DNSFILTER rules would have priority over OPENVPN Client
           rules (when client has DNS set to Exclusive mode).
  - FIXED: traffic routing from the router itself would fail when
           restarting the firewall while using an ovpn client with
           policy rules in effect.


380.68_4 (4-Oct-2017)
  - CHANGED: Updated dnsmasq to 2.78 (contains a number of
             security fixes).
  - FIXED: rstats could crash at start time in some situations.
  - FIXED: QOS Scheduler would revert back to sfq after you had
           re-enabled QOS while (fq_)codel was already selected.
  - FIXED: Missing tabs on the Parental Control page.
  - FIXED: Realtek port status wouldn't auto-refresh on the Sysinfo
           page.
  - FIXED: Incorrect sort by remaining time on the DHCP Lease page.
  - FIXED: Some LAN clients couldn't be added to the TOR redirected
           client list.


380.68_2 (12-Sept-2017)
  - FIXED: Some models would show the wrong menu options while in
           Repeater mode.
  - FIXED: USB modem page not displayed if WAN type was set to USB.
  - FIXED: CVE-2017-12754 security issue.
  - FIXED: Incorrect LAN ports order on Networkmap (RT-AC3200)
           (Asus bug)
  - FIXED: Extra OpenVPN CA not properly handled for OpenVPN
           clients 3, 4 and 5.
  - FIXED: Invalid txrate shown on Wireless Client page if client
           isn't authenticated yet


380.68 (18-Aug-2017)
  - IMPORTANT: due to major webui changes, you will need to
               either flush your browser cache, or force it
               to reload the page (shift-reload) the first time
               you access the webui after upgrading to 380.68.

  - NEW: Merged GPL 380_7743 binary blobs for the RT-N66U.
  - NEW: Backported Ethernet port status report on the Network Map
         from GPL 382.
  - NEW: Description field added to OpenVPN client configuration
  - NEW: Added missing hash types to ipset_arm (Patch by john9527)
  - NEW: Added hostname Busybox applet, used by some Entware packages
  - NEW: Added TPROXY netfilter target module (ARM only)
  - CHANGED: Switched webui menu generation code to GPL 382 code.
             This new code is easier for me to maintain.
  - CHANGED: Used webui menu icons from GPL 382.
  - CHANGED: Re-organized VPN pages, merging some together.
  - CHANGED: Reworked VPNStatus page, will now refresh itself every

release/src/router/tor/src/tools/tor-gencert.c
/* Copyright (c) 2007-2015, The Tor Project, Inc. */
/* See LICENSE for licensing information */

#include "orconfig.h"

#include <stdio.h>
#include <string.h>

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#include "compat.h"

/* Some versions of OpenSSL declare X509_STORE_CTX_set_verify_cb twice in
 * x509.h and x509_vfy.h. Suppress the GCC warning so we can build with
 * -Wredundant-decl. */
DISABLE_GCC_WARNING(redundant-decls)

#include <openssl/evp.h>
#include <openssl/pem.h>
#include <openssl/rsa.h>
#include <openssl/objects.h>
#include <openssl/obj_mac.h>
#include <openssl/err.h>

ENABLE_GCC_WARNING(redundant-decls)

#include <errno.h>
#if 0
#include <stdlib.h>
#include <stdarg.h>
#include <assert.h>
#endif

#include "compat.h"
#include "util.h"
#include "torlog.h"
#include "crypto.h"
#include "address.h"
#include "util_format.h"

#define IDENTITY_KEY_BITS 3072
#define SIGNING_KEY_BITS 2048
#define DEFAULT_LIFETIME 12

/* These globals are set via command line options. */
static char *identity_key_file = NULL;
static char *signing_key_file = NULL;
static char *certificate_file = NULL;
static int reuse_signing_key = 0;
static int verbose = 0;
static int make_new_id = 0;
static int months_lifetime = DEFAULT_LIFETIME;
static int passphrase_fd = -1;
static char *address = NULL;

static char *passphrase = NULL;
static size_t passphrase_len = 0;

static EVP_PKEY *identity_key = NULL;
static EVP_PKEY *signing_key = NULL;

/** Write a usage message for tor-gencert to stderr. */
static void
show_help(void)
{
  fprintf(stderr, "Syntax:\n"
          "tor-gencert [-h|--help] [-v] [-r|--reuse] [--create-identity-key]\n"
          "        [-i identity_key_file] [-s signing_key_file] "
          "[-c certificate_file]\n"
          "        [-m lifetime_in_months] [-a address:port] "
          "[--passphrase-fd <fd>]\n");
}

/* XXXX copied from crypto.c */
static void
crypto_log_errors(int severity, const char *doing)
{
  unsigned long err;
  const char *msg, *lib, *func;
  while ((err = ERR_get_error()) != 0) {
    msg = (const char*)ERR_reason_error_string(err);
    lib = (const char*)ERR_lib_error_string(err);
    func = (const char*)ERR_func_error_string(err);
    if (!msg) msg = "(null)";
    if (!lib) lib = "(null)";
    if (!func) func = "(null)";
    if (doing) {
      tor_log(severity, LD_CRYPTO, "crypto error while %s: %s (in %s:%s)",
              doing, msg, lib, func);
    } else {
      tor_log(severity, LD_CRYPTO, "crypto error: %s (in %s:%s)",
              msg, lib, func);
    }
  }
}

/** Read the passphrase from the passphrase fd. */
static int
load_passphrase(void)
{
  char *cp;
  char buf[1024]; /* "Ought to be enough for anybody." */
  memset(buf, 0, sizeof(buf)); /* should be needless */
  ssize_t n = read_all(passphrase_fd, buf, sizeof(buf), 0);
  if (n < 0) {
    log_err(LD_GENERAL, "Couldn't read from passphrase fd: %s",
            strerror(errno));
    return -1;
  }
  /* We'll take everything from the buffer except for optional terminating
   * newline. */
  cp = memchr(buf, '\n', n);
  if (cp == NULL) {
    passphrase_len = n;
  } else {
    passphrase_len = cp-buf;
  }
  passphrase = tor_strndup(buf, passphrase_len);
  memwipe(buf, 0, sizeof(buf));
  return 0;
}

static void
clear_passphrase(void)
{
  if (passphrase) {
    memwipe(passphrase, 0, passphrase_len);
    tor_free(passphrase);
  }
}

/** Read the command line options from <b>argc</b> and <b>argv</b>,
 * setting global option vars as needed.
 */
static int
parse_commandline(int argc, char **argv)
{
  int i;
  log_severity_list_t s;
  for (i = 1; i < argc; ++i) {
    if (!strcmp(argv[i], "--help") || !strcmp(argv[i], "-h")) {
      show_help();
      return 1;
    } else if (!strcmp(argv[i], "-i")) {
      if (i+1>=argc) {
        fprintf(stderr, "No argument to -i\n");
        return 1;
      }
      if (identity_key_file) {
        fprintf(stderr, "Duplicate values for -i\n");
        return -1;
      }
      identity_key_file = tor_strdup(argv[++i]);
    } else if (!strcmp(argv[i], "-s")) {
      if (i+1>=argc) {
        fprintf(stderr, "No argument to -s\n");
        return 1;
      }
      if (signing_key_file) {
        fprintf(stderr, "Duplicate values for -s\n");
        return -1;
      }
      signing_key_file = tor_strdup(argv[++i]);
    } else if (!strcmp(argv[i], "-c")) {
      if (i+1>=argc) {
        fprintf(stderr, "No argument to -c\n");
        return 1;
      }
      if (certificate_file) {
        fprintf(stderr, "Duplicate values for -c\n");
        return -1;
      }
      certificate_file = tor_strdup(argv[++i]);
    } else if (!strcmp(argv[i], "-m")) {
      if (i+1>=argc) {
        fprintf(stderr, "No argument to -m\n");
        return 1;
      }
      months_lifetime = atoi(argv[++i]);
      if (months_lifetime > 24 || months_lifetime < 0) {
        fprintf(stderr, "Lifetime (in months) was out of range.\n");
        return 1;
      }
    } else if (!strcmp(argv[i], "-r") || !strcmp(argv[i], "--reuse")) {
      reuse_signing_key = 1;
    } else if (!strcmp(argv[i], "-v")) {
      verbose = 1;
    } else if (!strcmp(argv[i], "-a")) {
      uint32_t addr;
      uint16_t port;
      char b[INET_NTOA_BUF_LEN];
      struct in_addr in;
      if (i+1>=argc) {
        fprintf(stderr, "No argument to -a\n");
        return 1;
      }
      if (addr_port_lookup(LOG_ERR, argv[++i], NULL, &addr, &port)<0)
        return 1;
      in.s_addr = htonl(addr);
      tor_inet_ntoa(&in, b, sizeof(b));
      tor_asprintf(&address, "%s:%d", b, (int)port);
    } else if (!strcmp(argv[i], "--create-identity-key")) {
      make_new_id = 1;
    } else if (!strcmp(argv[i], "--passphrase-fd")) {
      if (i+1>=argc) {
        fprintf(stderr, "No argument to --passphrase-fd\n");
        return 1;
      }
      passphrase_fd = atoi(argv[++i]);
    } else {
      fprintf(stderr, "Unrecognized option %s\n", argv[i]);
      return 1;
    }
  }

  memwipe(&s, 0, sizeof(s));
  if (verbose)
    set_log_severity_config(LOG_DEBUG, LOG_ERR, &s);
  else
    set_log_severity_config(LOG_WARN, LOG_ERR, &s);
  add_stream_log(&s, "<stderr>", fileno(stderr));

  if (!identity_key_file) {
    identity_key_file = tor_strdup("./authority_identity_key");
    log_info(LD_GENERAL, "No identity key file given; defaulting to %s",
             identity_key_file);
  }
  if (!signing_key_file) {
    signing_key_file = tor_strdup("./authority_signing_key");
    log_info(LD_GENERAL, "No signing key file given; defaulting to %s",
             signing_key_file);
  }
  if (!certificate_file) {
    certificate_file = tor_strdup("./authority_certificate");
    log_info(LD_GENERAL, "No signing key file given; defaulting to %s",
             certificate_file);
  }
  if (passphrase_fd >= 0) {
    if (load_passphrase()<0)
      return 1;
  }
  return 0;
}

static RSA *
generate_key(int bits)
{
  RSA *rsa = NULL;
  crypto_pk_t *env = crypto_pk_new();
  if (crypto_pk_generate_key_with_bits(env,bits)<0)
    goto done;
  rsa = crypto_pk_get_rsa_(env);
  rsa = RSAPrivateKey_dup(rsa);
 done:
  crypto_pk_free(env);
  return rsa;
}

/** Try to read the identity key from <b>identity_key_file</b>.  If no such
 * file exists and create_identity_key is set, make a new identity key and
 * store it.  Return 0 on success, nonzero on failure.
 */
static int
load_identity_key(void)
{
  file_status_t status = file_status(identity_key_file);
  FILE *f;

  if (make_new_id) {
    open_file_t *open_file = NULL;
    RSA *key;
    if (status != FN_NOENT) {
      log_err(LD_GENERAL, "--create-identity-key was specified, but %s "
              "already exists.", identity_key_file);
      return 1;
    }
    log_notice(LD_GENERAL, "Generating %d-bit RSA identity key.",
               IDENTITY_KEY_BITS);
    if (!(key = generate_key(IDENTITY_KEY_BITS))) {
      log_err(LD_GENERAL, "Couldn't generate identity key.");
      crypto_log_errors(LOG_ERR, "Generating identity key");
      return 1;
    }
    identity_key = EVP_PKEY_new();
    if (!(EVP_PKEY_assign_RSA(identity_key, key))) {
      log_err(LD_GENERAL, "Couldn't assign identity key.");
      return 1;
    }

    if (!(f = start_writing_to_stdio_file(identity_key_file,
                                          OPEN_FLAGS_REPLACE | O_TEXT, 0400,
                                          &open_file)))
      return 1;

    /* Write the key to the file.  If passphrase is not set, takes it from
     * the terminal. */
    if (!PEM_write_PKCS8PrivateKey_nid(f, identity_key,
                                       NID_pbe_WithSHA1And3_Key_TripleDES_CBC,
                                       passphrase, (int)passphrase_len,
                                       NULL, NULL)) {
      log_err(LD_GENERAL, "Couldn't write identity key to %s",
              identity_key_file);
      crypto_log_errors(LOG_ERR, "Writing identity key");
      abort_writing_to_file(open_file);
      return 1;
    }
    finish_writing_to_file(open_file);
  } else {
    if (status != FN_FILE) {
      log_err(LD_GENERAL,
              "No identity key found in %s.  To specify a location "
              "for an identity key, use -i.  To generate a new identity key, "
              "use --create-identity-key.", identity_key_file);
      return 1;
    }

    if (!(f = fopen(identity_key_file, "r"))) {
      log_err(LD_GENERAL, "Couldn't open %s for reading: %s",
              identity_key_file, strerror(errno));
      return 1;
    }

    /* Read the key.  If passphrase is not set, takes it from the terminal. */
    identity_key = PEM_read_PrivateKey(f, NULL, NULL, passphrase);
    if (!identity_key) {
      log_err(LD_GENERAL, "Couldn't read identity key from %s",
              identity_key_file);
      fclose(f);
      return 1;
    }
    fclose(f);
  }
  return 0;
}

/** Load a saved signing key from disk.  Return 0 on success, nonzero on
 * failure. */
static int
load_signing_key(void)
{
  FILE *f;
  if (!(f = fopen(signing_key_file, "r"))) {
    log_err(LD_GENERAL, "Couldn't open %s for reading: %s",
            signing_key_file, strerror(errno));
    return 1;
  }
  if (!(signing_key = PEM_read_PrivateKey(f, NULL, NULL, NULL))) {
    log_err(LD_GENERAL, "Couldn't read siging key from %s", signing_key_file);
    fclose(f);
    return 1;
  }
  fclose(f);
  return 0;
}

/** Generate a new signing key and write it to disk.  Return 0 on success,
 * nonzero on failure. */
static int
generate_signing_key(void)
{
  open_file_t *open_file;
  FILE *f;
  RSA *key;
  log_notice(LD_GENERAL, "Generating %d-bit RSA signing key.",
             SIGNING_KEY_BITS);
  if (!(key = generate_key(SIGNING_KEY_BITS))) {
    log_err(LD_GENERAL, "Couldn't generate signing key.");
    crypto_log_errors(LOG_ERR, "Generating signing key");
    return 1;
  }
  signing_key = EVP_PKEY_new();
  if (!(EVP_PKEY_assign_RSA(signing_key, key))) {
    log_err(LD_GENERAL, "Couldn't assign signing key.");
    return 1;
  }

  if (!(f = start_writing_to_stdio_file(signing_key_file,
                                        OPEN_FLAGS_REPLACE | O_TEXT, 0600,
                                        &open_file)))
    return 1;

  /* Write signing key with no encryption. */
  if (!PEM_write_RSAPrivateKey(f, key, NULL, NULL, 0, NULL, NULL)) {
    crypto_log_errors(LOG_WARN, "writing signing key");
    abort_writing_to_file(open_file);
    return 1;
  }

  finish_writing_to_file(open_file);

  return 0;
}

/** Encode <b>key</b> in the format used in directory documents; return
 * a newly allocated string holding the result or NULL on failure. */
static char *
key_to_string(EVP_PKEY *key)
{
  BUF_MEM *buf;
  BIO *b;
  RSA *rsa = EVP_PKEY_get1_RSA(key);
  char *result;
  if (!rsa)
    return NULL;

  b = BIO_new(BIO_s_mem());
  if (!PEM_write_bio_RSAPublicKey(b, rsa)) {
    crypto_log_errors(LOG_WARN, "writing public key to string");
    RSA_free(rsa);
    return NULL;
  }

  BIO_get_mem_ptr(b, &buf);
  result = tor_malloc(buf->length + 1);
  memcpy(result, buf->data, buf->length);
  result[buf->length] = 0;

  BIO_free(b);

  RSA_free(rsa);
  return result;
}

/** Set <b>out</b> to the hex-encoded fingerprint of <b>pkey</b>. */
static int
get_fingerprint(EVP_PKEY *pkey, char *out)
{
  int r = 1;
  crypto_pk_t *pk = crypto_new_pk_from_rsa_(EVP_PKEY_get1_RSA(pkey));
  if (pk) {
    r = crypto_pk_get_fingerprint(pk, out, 0);
    crypto_pk_free(pk);
  }
  return r;
}

/** Set <b>out</b> to the hex-encoded fingerprint of <b>pkey</b>. */
static int
get_digest(EVP_PKEY *pkey, char *out)
{
  int r = 1;
  crypto_pk_t *pk = crypto_new_pk_from_rsa_(EVP_PKEY_get1_RSA(pkey));
  if (pk) {
    r = crypto_pk_get_digest(pk, out);
    crypto_pk_free(pk);
  }
  return r;
}

/** Generate a new certificate for our loaded or generated keys, and write it
 * to disk.  Return 0 on success, nonzero on failure. */
static int
generate_certificate(void)
{
  char buf[8192];
  time_t now = time(NULL);
  struct tm tm;
  char published[ISO_TIME_LEN+1];
  char expires[ISO_TIME_LEN+1];
  char id_digest[DIGEST_LEN];
  char fingerprint[FINGERPRINT_LEN+1];
  char *ident = key_to_string(identity_key);
  char *signing = key_to_string(signing_key);
  FILE *f;
  size_t signed_len;
  char digest[DIGEST_LEN];
  char signature[1024]; /* handles up to 8192-bit keys. */
  int r;

  get_fingerprint(identity_key, fingerprint);
  get_digest(identity_key, id_digest);

  tor_localtime_r(&now, &tm);
  tm.tm_mon += months_lifetime;

  format_iso_time(published, now);
  format_iso_time(expires, mktime(&tm));

  tor_snprintf(buf, sizeof(buf),
               "dir-key-certificate-version 3"
               "%s%s"
               "\nfingerprint %s\n"
               "dir-key-published %s\n"
               "dir-key-expires %s\n"
               "dir-identity-key\n%s"
               "dir-signing-key\n%s"
               "dir-key-crosscert\n"
               "-----BEGIN ID SIGNATURE-----\n",
               address?"\ndir-address ":"", address?address:"",
               fingerprint, published, expires, ident, signing
               );
  tor_free(ident);
  tor_free(signing);

  /* Append a cross-certification */
  RSA *rsa = EVP_PKEY_get1_RSA(signing_key);
  r = RSA_private_encrypt(DIGEST_LEN, (unsigned char*)id_digest,
                          (unsigned char*)signature,
                          rsa,
                          RSA_PKCS1_PADDING);
  RSA_free(rsa);

  signed_len = strlen(buf);
  base64_encode(buf+signed_len, sizeof(buf)-signed_len, signature, r,
                BASE64_ENCODE_MULTILINE);

  strlcat(buf,
          "-----END ID SIGNATURE-----\n"
          "dir-key-certification\n", sizeof(buf));

  signed_len = strlen(buf);
  SHA1((const unsigned char*)buf,signed_len,(unsigned char*)digest);

  rsa = EVP_PKEY_get1_RSA(identity_key);
  r = RSA_private_encrypt(DIGEST_LEN, (unsigned char*)digest,
                          (unsigned char*)signature,
                          rsa,
                          RSA_PKCS1_PADDING);
  RSA_free(rsa);
  strlcat(buf, "-----BEGIN SIGNATURE-----\n", sizeof(buf));
  signed_len = strlen(buf);
  base64_encode(buf+signed_len, sizeof(buf)-signed_len, signature, r,
                BASE64_ENCODE_MULTILINE);
  strlcat(buf, "-----END SIGNATURE-----\n", sizeof(buf));

  if (!(f = fopen(certificate_file, "w"))) {
    log_err(LD_GENERAL, "Couldn't open %s for writing: %s",
            certificate_file, strerror(errno));
    return 1;
  }

  if (fputs(buf, f) < 0) {
    log_err(LD_GENERAL, "Couldn't write to %s: %s",
            certificate_file, strerror(errno));
    fclose(f);
    return 1;
  }
  fclose(f);
  return 0;
}

/** Entry point to tor-gencert */
int
main(int argc, char **argv)
{
  int r = 1;
  init_logging(1);

  /* Don't bother using acceleration. */
  if (crypto_global_init(0, NULL, NULL)) {
    fprintf(stderr, "Couldn't initialize crypto library.\n");
    return 1;
  }
  if (crypto_seed_rng()) {
    fprintf(stderr, "Couldn't seed RNG.\n");
    goto done;
  }
  /* Make sure that files are made private. */
  umask(0077);

  if (parse_commandline(argc, argv))
    goto done;
  if (load_identity_key())
    goto done;
  if (reuse_signing_key) {
    if (load_signing_key())
      goto done;
  } else {
    if (generate_signing_key())
      goto done;
  }
  if (generate_certificate())
    goto done;

  r = 0;
 done:
  clear_passphrase();
  if (identity_key)
    EVP_PKEY_free(identity_key);
  if (signing_key)
    EVP_PKEY_free(signing_key);
  tor_free(address);
  tor_free(identity_key_file);
  tor_free(signing_key_file);
  tor_free(certificate_file);
  tor_free(address);

  crypto_global_cleanup();
  return r;
}

